src/cargo/util/toml/mod.rs

```diff
@@ -773,6 +757,223 @@ impl<'de> de::Deserialize<'de> for VecStringOrBool {
+    where
+        D: de::Deserializer<'de>,
+    {
+        struct MaybeWorkspaceVisitor<T> {
+            marker: PhantomData<fn() -> MaybeWorkspace<T>>,
+        }
+
+        impl<'de, T> de::Visitor<'de> for MaybeWorkspaceVisitor<T>
+        where
+            T: de::Deserialize<'de>,
+        {
+            type Value = MaybeWorkspace<T>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("{ workspace: true } or a valid value")
+            }
+
+            fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                T::deserialize(v.into_deserializer()).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_str<E>(self, s: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                T::deserialize(s.into_deserializer()).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_i64<E>(self, numeric: i64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                T::deserialize(numeric.into_deserializer()).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_f64<E>(self, numeric: f64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                T::deserialize(numeric.into_deserializer()).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_u64<E>(self, numeric: u64) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                T::deserialize(numeric.into_deserializer()).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_seq<V>(self, seq: V) -> Result<Self::Value, V::Error>
+            where
+                V: de::SeqAccess<'de>,
+            {
+                let svd = de::value::SeqAccessDeserializer::new(seq);
+                T::deserialize(svd).map(MaybeWorkspace::Defined)
+            }
+
+            fn visit_map<V>(self, map: V) -> Result<Self::Value, V::Error>
+            where
+                V: de::MapAccess<'de>,
+            {
+                let mvd = de::value::MapAccessDeserializer::new(map);
+                TomlWorkspaceField::deserialize(mvd).and_then(|t| {
+                    if t.workspace {
+                        Ok(MaybeWorkspace::Workspace(true))
+                    } else {
+                        Err(de::Error::custom("workspace cannot be false"))
+                    }
+                })
+            }
+        }
+
+        deserializer.deserialize_any(MaybeWorkspaceVisitor {
+            marker: PhantomData,
+        })
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+enum MaybeWorkspaceBadge {
+    Workspace(TomlWorkspaceField),
+    Defined(BTreeMap<String, BTreeMap<String, String>>),
+}
+
+/// This exists only to provide a nicer error message.
```

Comment:
It would be nice to clean this up, but am not sure how to implement `Deserialized` for an untagged enum like this.